src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,50 +1,162 @@
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            linkLast(value);
+        } else {
+            Node<T> target = getNode(index);
+            linkBefore(value, target);
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        if (list.size() == 0) {
+            return false;
+        }
+        list.forEach(this::linkLast);
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> target = getNode(index);
+        T oldValue = target.value;
+        target.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> target = getNode(index);
+        unLink(target);
+        return target.value;
     }
 
     @Override
     public boolean remove(T t) {
+        for (Node<T> node = first; node != null; node = node.next) {
+            if (Objects.equals(node.value, t)) {
+                unLink(node);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void linkLast(T value) {
+        Node<T> oldLast = last;
+        last = new Node<>(oldLast, value, null);
+        if (oldLast == null) {
+            first = last;
+        } else {
+            oldLast.next = last;
+        }
+        size++;
+    }
+
+    private void linkBefore(T value, Node<T> target) {
+        Node<T> previous = target.prev;
+        Node<T> newNode = new Node<>(previous, value, target);
+        target.prev = newNode;
+        if (previous == null) {
+            first = newNode;
+        } else {
+            previous.next = newNode;
+        }
+        size++;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
+    }
+
+    private Node<T> getNode(int index) {
+        checkIndex(index);
+        if (index < size >> 1) {
+            Node<T> node = first;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+            return node;
+        } else {
+            Node<T> node = last;
+            for (int i = size - 1; i > index; i--) {
+                node = node.prev;
+            }
+            return node;
+        }
+    }
+
+    private void unLink(Node<T> node) {
+        Node<T> previous = node.prev;
+        Node<T> next = node.next;
+
+        if (previous == null) {
+            first = next;
+        } else {
+            previous.next = next;
```

Comment:
```suggestion
            previous.next = next;
            node.prev = null;
```
in this task it is important to understand the concept of links
so let`s remove the links from both sides